server/webex/client.go

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package webex
+
+import (
+	"bytes"
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"github.com/pkg/errors"
+)
+
+const StatusStarted = "STARTED"
+
+type ClientError struct {
+	StatusCode int
+	Err        error
+}
+
+func (ce *ClientError) Error() string {
+	return ce.Err.Error()
+}
+
+type Client interface {
+	GetPersonalMeetingRoomUrl(roomId, username, email string) (string, *ClientError)
+}
+
+// Client represents a Webex API client
+type client struct {
+	httpClient *http.Client
+	xmlURL     string
+	siteName   string
+}
+
+// NewClient returns a new Webex XML API client.
+func NewClient(siteHost, siteName string) Client {
+	webexURL := (&url.URL{
+		Scheme: "https",
+		Host:   siteHost,
+		Path:   "/WBXService/XMLService",
+	}).String()
+
+	return &client{
+		httpClient: &http.Client{},
+		xmlURL:     webexURL,
+		siteName:   siteName,
+	}
+}
+
+// GetPersonalMeetingRoomUrl prefers roomId, username, and email for finding the PMR url (in that order).
+func (c *client) GetPersonalMeetingRoomUrl(roomId, username, email string) (string, *ClientError) {
+	if roomId != "" {
+		pmrUrl, err := c.getPMRFromRoomId(roomId)
+		if err == nil && pmrUrl != "" {
+			return pmrUrl, nil
+		}
+	}
+	if username != "" {
+		pmrUrl, err := c.getPMRFromUserName(username)
+		if err == nil && pmrUrl != "" {
+			return pmrUrl, nil
+		}
+	}
+	if email != "" {
+		pmrUrl, err := c.getPMRFromEmail(email)
+		if err == nil && pmrUrl != "" {
+			return pmrUrl, nil
+		}
+	}
+
+	return "", &ClientError{500, errors.New("couldn't get PMR url")}
+}
+
+const payloadWrapper = `<?xml version="1.0" encoding="UTF-8"?>
```

Comment:
Have you used https://golang.org/pkg/text/template/ before? It's a bit clearer way to do this kind of thing. 0/5 this is simple enough that it's not a big deal.

Reply:
@crspeller I haven't personally, but of course I've seen it used in the Jira plugin. Personally, I think it's a little heavyweight (I'm surprised it's so complicated actually...)